### PR TITLE
Initial support for matrix inverse in qdk_sim_experimental crate

### PR DESCRIPTION
**Pull request by cgranade**
This PR begins work on #915 by adding a pure-Rust implementation of LU decompositions and matrix inverses, drawing from the MIT-licensed examples at https://github.com/cgranade/cursed-linalg. These linear algebra routines will be needed in order to implement `expm` without passing a LAPACK dependency through to end users.

The PR will appear very messy until #979 is merged, as it includes unrelated changes from `main`.

Note that this PR targets a feature branch, and further work on unit tests, integration tests, and perf benchmarks are expected before that feature is merged to main.

### CODE REVIEW

**Review by kuzminrobin — APPROVED** [review]
I'm learning Rust and still don't understand most of the code here. But at a glance I didn't see any thing to report. 
Approving to unblock you, but I would recommend an approval from someone more knowledgeable than me.

**Comment by cgranade** [comment]
@kuzminrobin: Thanks! If you're curious about any of the Rust-y parts, I'd be happy to sync with you and go over in more detail.

In the meantime, @swernli and @samarsha, if you have any comments on Rust-specific points I'd greatly appreciate your feedback. Thanks!

File: src/Simulation/qdk_sim_rs/src/linalg/array_ext.rs
Comment (samarsha) [inline src/Simulation/qdk_sim_rs/src/linalg/array_ext.rs:20]: So for extension traits there's [RFC 445](https://rust-lang.github.io/rfcs/0445-extension-trait-conventions.html). It looks like the traits in this module don't follow those naming conventions.

File: src/Simulation/qdk_sim_rs/src/linalg/array_ext.rs
Comment (samarsha) [inline src/Simulation/qdk_sim_rs/src/linalg/array_ext.rs:22]: This is a very general error type but there is only one possible error in this case. Maybe `Option<usize>`?

File: src/Simulation/qdk_sim_rs/src/linalg/decompositions/lu.rs
Comment (samarsha) [inline src/Simulation/qdk_sim_rs/src/linalg/decompositions/lu.rs:29]: Are `LUDecompasable` and `LUDecomposition` also extension traits? I only see an impl for one type each. Is it needed for crate consumers to write their own impls?

**Review by samarsha — COMMENTED** [review]
I've only looked through half of the files so far.

File: src/Simulation/qdk_sim_rs/README.md
Comment (cgranade) [reply src/Simulation/qdk_sim_rs/README.md:147]: Yes, the rest is filled in by the `document-features` crate based on documentation comments in `Cargo.toml` itself.

File: src/Simulation/qdk_sim_rs/src/linalg/array_ext.rs
Comment (cgranade) [reply src/Simulation/qdk_sim_rs/src/linalg/array_ext.rs:20]: Ah, that it should be `ArrayBaseExt`?

File: src/Simulation/qdk_sim_rs/src/linalg/decompositions/lu.rs
Comment (cgranade) [reply src/Simulation/qdk_sim_rs/src/linalg/decompositions/lu.rs:29]: They're note extension traits, no, but an attempt to generalize 𝐿𝑈 decompositions to allow for application over other kinds of scalars and data structures including precomputed decompositions.

File: src/Simulation/qdk_sim_rs/README.md
Comment (samarsha) [reply src/Simulation/qdk_sim_rs/README.md:147]: I see, I assume it's appending it to the end of the file (rather than looking for a specific header name)? A comment in the readme could help avoid accidentally deleting the section or adding sections after it.

**Review by swernli — APPROVED** [review]
Some minor feedback below, but overall it looks great!

File: src/Simulation/qdk_sim_rs/README.md
Comment (cgranade) [reply src/Simulation/qdk_sim_rs/README.md:147]: That's right, yeah. There's two `#![cfg_attr(..., doc = ...)]` attributes at https://github.com/microsoft/qsharp-runtime/pull/920/files#diff-09741599828fb3d312b03bab5767da1f994d179496ea35a281db89a47339aebfR4, which causes rustdoc to append the contents of each attribute together. In particular, the first `#![doc(...)]` attribute uses `include_str` to bring in README.md, while the second one calls `document_features`.

Will add a note to that effect, then, as you suggest.

File: src/Simulation/qdk_sim_rs/src/linalg/array_ext.rs
Comment (samarsha) [reply src/Simulation/qdk_sim_rs/src/linalg/array_ext.rs:20]: Yeah, I think so.

- `ShapeExt` -> `ArrayBaseExt`
- `MatrixExt` -> `ArrayBaseMatrixExt`
- `RemoveAxisExt` -> `ArrayBaseRemoveAxisExt`

(Of course, assuming that there will never be any other impls for these - for example, n-d arrays represented by something other than an ArrayBase. Extension traits are kind of weird since the line between a single-impl trait and a many-impl trait seems blurry.)

File: src/Simulation/qdk_sim_rs/src/linalg/decompositions/lu.rs
Comment (samarsha) [reply src/Simulation/qdk_sim_rs/src/linalg/decompositions/lu.rs:29]: If it's a generalization, that would imply there should be multiple impls for each trait, right? Are you planning to add some later?

File: src/Simulation/qdk_sim_rs/src/tableau.rs
Comment (cgranade) [reply src/Simulation/qdk_sim_rs/src/tableau.rs:105]: My goal here had been to offer more specific error types before stabilizing the API, but at this point it wasn't clear if there should be a general-purpose assertion error, or one more focused to the `actual != expected` format.

File: src/Simulation/qdk_sim_rs/src/utils.rs
Comment (cgranade) [reply src/Simulation/qdk_sim_rs/src/utils.rs:44]: The logging API as written only supports returning strings as errors; I suppose that's as good a reason as any to keep a `MiscError` type? Maybe I could weaken the comment to something like "consider more specific types if applicable"?

File: src/Simulation/qdk_sim_rs/src/linalg/array_ext.rs
Comment (cgranade) [reply src/Simulation/qdk_sim_rs/src/linalg/array_ext.rs:20]: Sounds good, I'll go on and modify, then.

File: src/Simulation/qdk_sim_rs/src/linalg/decompositions/lu.rs
Comment (cgranade) [reply src/Simulation/qdk_sim_rs/src/linalg/decompositions/lu.rs:29]: Not necessarily, but the corresponding split for eigendecompositions has proved useful (allows for precomputed decompositions as well, for example) such that leaving it open for consumers to write their own implementations may be useful as well. That does to some level presuppose using the linear algebra functionality outside of the context of the `qdk_sim` crate, but ideally I'd love to split that out into its own crate.

File: src/Simulation/qdk_sim_rs/Cargo.toml
Comment (cgranade) [reply src/Simulation/qdk_sim_rs/Cargo.toml]: Good catch, thanks! Fixed.

File: src/Simulation/qdk_sim_rs/src/linalg/array_ext.rs
Comment (cgranade) [reply src/Simulation/qdk_sim_rs/src/linalg/array_ext.rs:90]: Implicitly in functions which use it, which in theory should work since it's not a public API, but I'll go on and add a test case to be sure. Thanks for the suggestion!

File: src/Simulation/qdk_sim_rs/src/linalg/decompositions/lu.rs
Comment (samarsha) [reply src/Simulation/qdk_sim_rs/src/linalg/decompositions/lu.rs:29]: Ok, sounds good, thanks for explaining.

Comment (cgranade) [reply src/Simulation/qdk_sim_rs/src/linalg/decompositions/lu.rs:29]: Absolutely, thanks for the comment! ♥

File: src/Simulation/qdk_sim_rs/src/linalg/array_ext.rs
Comment (cgranade) [reply src/Simulation/qdk_sim_rs/src/linalg/array_ext.rs:22]: My thinking here was that this is an internal-only trait/impl pair (`pub(crate)` in particular), and is mainly used to generate the specific error case `NotSquare` as a utility. Were this a public API, I'd agree entirely, but making that change here would necessitate adding a bunch of `.ok_or` calls where the return value for this extension method is used.

File: src/Simulation/qdk_sim_rs/src/utils.rs
Comment (cgranade) [reply src/Simulation/qdk_sim_rs/src/utils.rs:44]: Modified the comment on `MiscError` to be a bit less agressive. Thanks!

**Comment by cgranade** [comment]
@samarsha: Wanted to double-check that all your feedback has been addressed before merging. Thanks!

**Comment by samarsha** [comment]
Yes, looks good, thanks.

**Comment by cgranade** [comment]
Awesome, thank you for all your detailed feedback! 💕

**Pull request merged**
